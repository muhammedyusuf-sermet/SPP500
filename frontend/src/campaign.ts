export interface ICampaignState {
	Id?: string,
	Name: string,
	Description: string;
	// Monsters: Monster.IMonsterState[];
}
