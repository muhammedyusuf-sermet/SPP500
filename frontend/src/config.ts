export const API_URL = "https://http://3.18.65.138:3000";