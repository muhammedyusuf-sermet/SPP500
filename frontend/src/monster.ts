export enum Size {
	Tiny = "Tiny",
	Small = "Small",
	Medium = "Medium",
	Large = "Large",
	Huge = "Huge",
	Gargantuan = "Gargantuan"
}

export enum MonsterType {
	Beast = "Beast",
	Dragon = "Dragon",
	Monstrosity = "Monstrosity",
	Humanoid = "Humanoid",
	Fiend = "Fiend",
	Undead = "Undead",
	Giant = "Giant",
	Elemental = "Elemental",
	SwarmOfTinyBeasts = "SwarmOfTinyBeasts",
	Construct = "Construct",
	Celestial = "Celestial",
	Aberration = "Aberration",
	Fey = "Fey",
	Plant = "Plant",
	Ooze = "Ooze"
}

export enum MonsterRace {
	AnyRace = "AnyRace",
	Devil = "Devil",
	Demon = "Demon",
	Human = "Human",
	Shapechanger = "Shapechanger",
	Goblinoid = "Goblinoid",
	Titan = "Titan",
	Gnoll = "Gnoll",
	Gnome = "Gnome",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Orc = "Orc",
	Kobold = "Kobold",
	Lizardfolk = "Lizardfolk",
	Merfolk = "Merfolk",
	Sahuagin = "Sahuagin",
	Grimlock = "Grimlock"
}

export enum Alignment {
	Unaligned = "Unaligned",
	AnyAlignment = "AnyAlignment",
	AnyNonGoodAlignment = "AnyNonGoodAlignment",
	AnyNonEvilAlignment = "AnyNonEvilAlignment",
	AnyNonLawfulAlignment = "AnyNonLawfulAlignment",
	AnyNonChaoticAlignment = "AnyNonChaoticAlignment",
	AnyGoodAlignment = "AnyGoodAlignment",
	AnyEvilAlignment = "AnyEvilAlignment",
	AnyLawfulAlignment = "AnyLawfulAlignment",
	AnyChaoticAlignment = "AnyChaoticAlignment",
	AnyNeutralAlignment = "AnyNeutralAlignment",
	LawfulGood = "LawfulGood",
	LawfulNeutral = "LawfulNeutral",
	LawfulEvil = "LawfulEvil",
	NeutralGood = "NeutralGood",
	Neutral = "Neutral",
	NeutralEvil = "NeutralEvil",
	ChaoticGood = "ChaoticGood",
	ChaoticNeutral = "ChaoticNeutral",
	ChaoticEvil = "ChaoticEvil"
}

export enum Environment {
	Arctic = "Arctic",
	Coastal = "Coastal",
	Desert = "Desert",
	Forest = "Forest",
	Grassland = "Grassland",
	Hill = "Hill",
	Mountain = "Mountain",
	Swamp = "Swamp",
	Underdark = "Underdark",
	Underwater = "Underwater",
	Urban = "Urban"
}

export enum MonsterAction {
	SpecialAbility = "SpecialAbility",
	Action = "Action",
	LegendaryAction = "LegendaryAction"
}

export interface IMonsterAbilities {
	name: string,
	description: string,
}

export enum TypeOfAction {
	Ranged = "Ranged",
	Melee = "Melee",
}

export interface IMonsterActions {
	name: string,
	description: string,
	usageCount: number,
	legendaryAction: boolean,
	typeOfAction?: TypeOfAction,
	toHit: number,
	reach: number,
	targets: number,
	averagedDamage: number,
	damageDice: number,
	damageConstant: number,

}

export interface IMonsterState {
	Id?: number,
	Name: string,
	Size?: Size | string;
	Type?: MonsterType | string;
	Race?: MonsterRace | string;
	Environment?: Environment | string;
	Alignment?: Alignment | string;

	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;

	Speed?: string;
	Languages?: string;

	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;

	ChallengeRating?: number;

	AbilityScores: IMonsterAbilityScoreState;
	Senses: SenseMap;
	Skills: SkillMap;
	SavingThrows: IMonsterSavingThrowState;
	//Actions: IActionState[];
}

export interface SkillMap {
	[skillName: string]: number|undefined;
}

export interface SenseMap {
	[senseName: string]: number|undefined;
}

export interface IMonsterAbilityScoreState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterSavingThrowState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IActionState {
	Name: string,
	Description: string,
	HitBonus?: number,
	Damage?: string,
	DamageBonus?: number,
	Type?: MonsterAction
}
/*export interface IMonster {
	name: string,
	type: MonsterType,
	alignment: MonsterAlignment,
	size: MonsterSize,
	race: MonsterRace,
	environment: MonsterEnvironment,
	resistance?: string,
	damageImmunity?: string,
	conditionImmunity?: string,
	vulnerability?: string,
	armorClass: number,
	hitPoints?: number,
	hitPointDice?: string,
	hitPointDiceAdd?: number,
	speedLand: number,
	speedSwim?: number,
	strStat: number,
	dexStat: number,
	conStat: number,
	intStat: number,
	wisStat: number,
	chaStat: number,
	strSavingThrow?: number,
	dexSavingThrow?: number,
	conSavingThrow?: number,
	intSavingThrow?: number,
	wisSavingThrow?: number,
	chaSavingThrow?: number,
	skillsAthletics: number,
	skillsAcrobatics?: number,
	skillsSleightOfHand?: number,
	skillsStealth?: number,
	skillsArcana?: number,
	skillsHistory?: number,
	skillsInvestigation?: number,
	skillsNature?: number,
	skillsReligion?: number,
	skillsAnimalHandling?: number,
	skillsInsight?: number,
	skillsMedicine?: number,
	skillsPerception?: number,
	skillsSurvival?: number,
	skillsDeception?: number,
	skillsIntimidation?: number,
	skillsPerformance?: number,
	skillsPersuasion?: number,
	sensesBlindsight?: number,
	sensesDarkvision?: number,
	sensesTremorsense?: number,
	sensesTruesight?: number,
	sensesPassivePerception?: number,
	sensesPassiveInvestigation?: number,
	sensesPassiveInsight?: number,
	languages?: string,
	challengeRating: number,
	experiencePoints: number,
	abilities?: IMonsterAbilities[],
	actions?: IMonsterActions[],
}*/