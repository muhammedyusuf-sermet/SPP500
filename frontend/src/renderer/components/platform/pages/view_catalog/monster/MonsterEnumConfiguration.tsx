import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { MonsterType, MonsterRace, Size, Environment, Alignment } from '../../../../../../monster';
import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid, Select, Tooltip } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import {Help} from "bloomer";

const types = Object.values(MonsterType);
const sizes = Object.values(Size);
const races = Object.values(MonsterRace);
const environments = Object.values(Environment);
const alignments = Object.values(Alignment);

export interface IMonsterDropdownProps {
	disabled?: boolean,
	name: string,
	selected?: string,
	options: string[]
	onChange: (name: string, selectOption?: string) => void
}

export interface IMonsterDropdownState {
	selected: string
}

export class MonsterDropdown extends React.Component<IMonsterDropdownProps, IMonsterDropdownState> {
	constructor(props: IMonsterDropdownProps) {
		super(props);
		this.state = {
			selected: props.selected ? props.selected : 'Default'
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDropdownProps) {
		if (this.state.selected != nextProps.selected)
			this.setState({
				selected: nextProps.selected ? nextProps.selected : 'Default'
			});
	}

	handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
		this.setState({selected: event.target.value});
		this.props.onChange(this.props.name, event.target.value);
	}

	render() {
		return (
			<Select
				native={true}
				id={this.props.name}
				disabled={this.props.disabled}
				onChange={this.handleChange}
				value={this.state.selected}
				input={<Input name={this.props.name} id={this.props.name} />} >
				<option
					id={this.props.name+'.'+'Default'}
					value='' >
					Default
				</option>
				{this.props.options.map(option =>
					<option
						key={this.props.name+'.'+option}
						id={this.props.name+'.'+option}
						value={option}>
						{option.replace(/([A-Z])/g, ' $1').trim()}
					</option>
				)}
			</Select>
		);
	}
}

export interface IMonsterEnumConfigurationProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Size?: string;
		Type?: string;
		Race?: string;
		Alignment?: string;
		Environment?: string;
	}
}

export interface IMonsterEnumConfigurationState {
	// data
	Size?: string;
	Type?: string;
	Race?: string;
	Alignment?: string;
	Environment?: string;
	// errors
	SizeError?: string;
	TypeError?: string;
	RaceError?: string;
	AlignmentError?: string;
	EnvironmentError?: string;
}

export class MonsterEnumConfiguration extends React.Component<IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState> {
	constructor(props: IMonsterEnumConfigurationProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterEnumConfigurationProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial
			});
	}

	handleMonsterEnumChange = (name: string, newEnum?: string) => {
		if (newEnum) {
			Joi.validate(
				newEnum,
				Joi.reach(this.props.PayloadSchema, [name]),
				this.props.ValidationOptions,
				(errors: ValidationError) => {
					if(errors) {
						this.setState({
							[name+'Error']: errors.details[0].message
						});
					} else {
						this.setState({
							[name]: newEnum,
							[name+'Error']: undefined
						});
					}
				});
		} else {
			this.setState({
				[name]: undefined,
				[name+'Error']: undefined
			});
		}
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Basic Configurations</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Type' >Type</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Type}
									name='Type'
									options={types}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Type-helper-text' >{this.state.TypeError}</FormHelperText>
								<Tooltip disableFocusListener title={"Creature types are rough categories of creatures which determine the way game mechanics affect the creature. The choice of type is important, as all creatures which have a given type will share certain characteristics (with some exceptions)."}>
										<Help>Hover for More About Monster Type</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Size' >Size</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Size}
									name='Size'
									options={sizes}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Size-helper-text' >{this.state.SizeError}</FormHelperText>
								<Tooltip disableFocusListener title={"A creature's size determines how much space it occupies on squares or hexes, as shown in the Creature Size and Space table. If the miniature you use for a monster takes up an amount of space different from what's on the table, that's fine, but treat the monster as its official size for all other rules."}>
										<Help>Hover for More About Monster Size</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Race' >Race</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Race}
									name='Race'
									options={races}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Race-helper-text' >{this.state.RaceError}</FormHelperText>
								<Tooltip disableFocusListener title={"A character race is a fundamental part of the identity and nature of characters in the Dungeons & Dragons role-playing game. Each race has a distinct appearance, behavior and often range of statistics associated with it. See your DM handbook for more information."}>
										<Help>Hover for More About Monster Race</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Alignment' >Alignment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Alignment}
									name='Alignment'
									options={alignments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Alignment-helper-text' >{this.state.AlignmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Alignment is a categorization of the ethical and moral perspective of player characters, non-player characters, and creatures. Most versions of the game feature a system in which players make two choices for characters. One is the character's views on 'law' vs 'chaos', the other on 'good' vs 'evil'."}>
										<Help>Hover for More About Monster Alignment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Environment' >Environment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Environment}
									name='Environment'
									options={environments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Environment-helper-text' >{this.state.EnvironmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Monster environment is the type of environment that the monster is most likely to be found in."}>
										<Help>Hover for More About Monster Environment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}