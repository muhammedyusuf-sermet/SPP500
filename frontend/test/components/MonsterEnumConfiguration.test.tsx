import * as React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';

import { MonsterType, Alignment, Size, MonsterRace, Environment } from "../../src/monster";
import { IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState, MonsterEnumConfiguration } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterEnumConfiguration';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

const payloadSchema = Joi.object({
	Size: Joi.string().valid(Joi.ref('$SizeOptions')),
	Type: Joi.string().valid(Joi.ref('$TypeOptions')),
	Race: Joi.string().valid(Joi.ref('$RaceOptions')),
	Alignment: Joi.string().valid(Joi.ref('$AlignmentOptions')),
	Environment: Joi.string().valid(Joi.ref('$EnvironmentOptions')),
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {
		SizeOptions: Object.keys(Size),
		TypeOptions: Object.keys(MonsterType),
		RaceOptions: Object.keys(MonsterRace),
		AlignmentOptions: Object.keys(Alignment),
		EnvironmentOptions: Object.keys(Environment)
	}
};

////// Happy Path //////
describe('Monster Creation', () => {

	let monsterEnumConfigurationInstance: ReactWrapper<IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState, MonsterEnumConfiguration>;

	describe('Provided Enum Configuration values', () => {

		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<MonsterEnumConfiguration, IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState>(
					<MonsterEnumConfiguration
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							Size: Size.Gargantuan,
							Type: MonsterType.Elemental,
							Race: MonsterRace.Dwarf,
							Alignment: Alignment.AnyEvilAlignment,
							Environment: Environment.Coastal
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('should have the selected Type', () => {
			expect(monsterEnumConfigurationInstance.find('select#Type').props().value).toEqual('Elemental');
		});

		it('should have the selected Size', () => {
			expect(monsterEnumConfigurationInstance.find('select#Size').props().value).toEqual('Gargantuan');
		});

		it('should have the selected Race', () => {
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Dwarf');
		});

		it('should have the selected Environment', () => {
			expect(monsterEnumConfigurationInstance.find('select#Environment').props().value).toEqual('Coastal');
		});

		it('should have the selected Alignment', () => {
			expect(monsterEnumConfigurationInstance.find('select#Alignment').props().value).toEqual('AnyEvilAlignment');
		});
	});

	describe('componentWillReceiveProps()', () => {
		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<MonsterEnumConfiguration, IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState>(
					<MonsterEnumConfiguration
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							Size: Size.Gargantuan,
							Type: MonsterType.Elemental,
							Race: MonsterRace.Dwarf,
							Alignment: Alignment.AnyEvilAlignment,
							Environment: Environment.Coastal
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('renders new props', () => {
			monsterEnumConfigurationInstance.setProps({
				initial:  {
					Size: Size.Tiny,
					Type: MonsterType.Construct,
					Race: MonsterRace.Gnome,
					Alignment: Alignment.ChaoticNeutral,
					Environment: Environment.Forest
				}
			})
			expect(monsterEnumConfigurationInstance.find('select#Type').props().value).toEqual('Construct');
			expect(monsterEnumConfigurationInstance.find('select#Size').props().value).toEqual('Tiny');
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Gnome');
			expect(monsterEnumConfigurationInstance.find('select#Environment').props().value).toEqual('Forest');
			expect(monsterEnumConfigurationInstance.find('select#Alignment').props().value).toEqual('ChaoticNeutral');
		});

		it('renders without crashing with same props', () => {
			monsterEnumConfigurationInstance.setProps({
				initial: {
					Size: Size.Gargantuan,
					Type: MonsterType.Elemental,
					Race: MonsterRace.Dwarf,
					Alignment: Alignment.AnyEvilAlignment,
					Environment: Environment.Coastal
				}
			});
			expect(monsterEnumConfigurationInstance.find('select#Type').props().value).toEqual('Elemental');
			expect(monsterEnumConfigurationInstance.find('select#Size').props().value).toEqual('Gargantuan');
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Dwarf');
			expect(monsterEnumConfigurationInstance.find('select#Environment').props().value).toEqual('Coastal');
			expect(monsterEnumConfigurationInstance.find('select#Alignment').props().value).toEqual('AnyEvilAlignment');
		});

		it('disables input if disable prop changes', () => {
			expect(monsterEnumConfigurationInstance.find('select#Type').props().disabled).toEqual(false);
			expect(monsterEnumConfigurationInstance.find('select#Size').props().disabled).toEqual(false);
			expect(monsterEnumConfigurationInstance.find('select#Race').props().disabled).toEqual(false);
			expect(monsterEnumConfigurationInstance.find('select#Environment').props().disabled).toEqual(false);
			expect(monsterEnumConfigurationInstance.find('select#Alignment').props().disabled).toEqual(false);
			monsterEnumConfigurationInstance.setProps({
				disabled: true
			})
			expect(monsterEnumConfigurationInstance.find('select#Type').props().disabled).toEqual(true);
			expect(monsterEnumConfigurationInstance.find('select#Size').props().disabled).toEqual(true);
			expect(monsterEnumConfigurationInstance.find('select#Race').props().disabled).toEqual(true);
			expect(monsterEnumConfigurationInstance.find('select#Environment').props().disabled).toEqual(true);
			expect(monsterEnumConfigurationInstance.find('select#Alignment').props().disabled).toEqual(true);
		});
	});

	describe('Small snapshot', () => {
		it ('matches snapshot', () => {
			const shallowMonsterEnumConfigurationInstance =
				shallow<MonsterEnumConfiguration, IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState>(
					<MonsterEnumConfiguration
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}}/>
				);
			expect(shallowMonsterEnumConfigurationInstance).toMatchSnapshot();
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<MonsterEnumConfiguration, IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState>(
					<MonsterEnumConfiguration
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterEnumConfigurationInstance.find('select#Type').simulate('change', { target: { value: "Celestial" } })
			monsterEnumConfigurationInstance.find('select#Size').simulate('change', { target: { value: "Gargantuan" } })
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: "Devil" } })
			monsterEnumConfigurationInstance.find('select#Environment').simulate('change', { target: { value: "Underdark" } })
			monsterEnumConfigurationInstance.find('select#Alignment').simulate('change', { target: { value: "AnyGoodAlignment" } })
			expect(monsterEnumConfigurationInstance.state()).toEqual({
				// data
				Size: 'Gargantuan',
				Type: 'Celestial',
				Race: 'Devil',
				Alignment: 'AnyGoodAlignment',
				Environment: 'Underdark',
				// errors
				SizeError: undefined,
				TypeError: undefined,
				RaceError: undefined,
				AlignmentError: undefined,
				EnvironmentError: undefined
			});
		});
		// STATE CHANGE
		it('should change only types when type is changed', () => {
			monsterEnumConfigurationInstance.find('select#Type').simulate('change', { target: { value: "Undead" } })
			expect(monsterEnumConfigurationInstance.state().Type).toEqual(MonsterType.Undead);
		})

		it('should change only alignment when alignment is changed', () => {
			monsterEnumConfigurationInstance.find('select#Alignment').simulate('change', { target: { value: "AnyNonEvilAlignment" } })
			expect(monsterEnumConfigurationInstance.state().Alignment).toEqual(Alignment.AnyNonEvilAlignment);
		})

		it('should change only size when size is changed', () => {
			monsterEnumConfigurationInstance.find('select#Size').simulate('change', { target: { value: "Large" } })
			expect(monsterEnumConfigurationInstance.state().Size).toEqual(Size.Large);
		})

		it('should change only races when race is changed', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: "Goblinoid" } })
			expect(monsterEnumConfigurationInstance.state().Race).toEqual(MonsterRace.Goblinoid);
		})

		it('should change only environments when environment is changed', () => {
			monsterEnumConfigurationInstance.find('select#Environment').simulate('change', { target: { value: "Coastal" } })
			expect(monsterEnumConfigurationInstance.state().Environment).toEqual(Environment.Coastal);
		})
		// DEFAULTS
		it('should change Type back to undefined when Default', () => {
			monsterEnumConfigurationInstance.find('select#Type').simulate('change', { target: { value: "Default" } })
			expect(monsterEnumConfigurationInstance.state().Type).toEqual(undefined);
		})

		it('should change Alignment back to undefined when Default', () => {
			monsterEnumConfigurationInstance.find('select#Alignment').simulate('change', { target: { value: "Default" } })
			expect(monsterEnumConfigurationInstance.state().Alignment).toEqual(undefined);
		})

		it('should change Size back to undefined when Default', () => {
			monsterEnumConfigurationInstance.find('select#Size').simulate('change', { target: { value: "Default" } })
			expect(monsterEnumConfigurationInstance.state().Size).toEqual(undefined);
		})

		it('should change Race back to undefined when Default', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: "Default" } })
			expect(monsterEnumConfigurationInstance.state().Race).toEqual(undefined);
		})

		it('should change Environment back to undefined when Default', () => {
			monsterEnumConfigurationInstance.find('select#Environment').simulate('change', { target: { value: "Default" } })
			expect(monsterEnumConfigurationInstance.state().Environment).toEqual(undefined);
		})
		// VALIDATION
		it('should validate types when type is changed', () => {
			monsterEnumConfigurationInstance.find('select#Type').simulate('change', { target: { value: "Invalid" } })
			expect(monsterEnumConfigurationInstance.state().Type).toEqual(undefined);
			expect(monsterEnumConfigurationInstance.state().TypeError).toEqual('\"value\" must be one of [context:TypeOptions]');
		})

		it('should validate alignment when alignment is changed', () => {
			monsterEnumConfigurationInstance.find('select#Alignment').simulate('change', { target: { value: "Invalid" } })
			expect(monsterEnumConfigurationInstance.state().Alignment).toEqual(undefined);
			expect(monsterEnumConfigurationInstance.state().AlignmentError).toEqual('\"value\" must be one of [context:AlignmentOptions]');
		})

		it('should validate size when size is changed', () => {
			monsterEnumConfigurationInstance.find('select#Size').simulate('change', { target: { value: "Invalid" } })
			expect(monsterEnumConfigurationInstance.state().Size).toEqual(undefined);
			expect(monsterEnumConfigurationInstance.state().SizeError).toEqual('\"value\" must be one of [context:SizeOptions]');
		})

		it('should validate races when race is changed', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: "Invalid" } })
			expect(monsterEnumConfigurationInstance.state().Race).toEqual(undefined);
			expect(monsterEnumConfigurationInstance.state().RaceError).toEqual('\"value\" must be one of [context:RaceOptions]');
		})

		it('should validate environments when environment is changed', () => {
			monsterEnumConfigurationInstance.find('select#Environment').simulate('change', { target: { value: "Invalid" } })
			expect(monsterEnumConfigurationInstance.state().Environment).toEqual(undefined);
			expect(monsterEnumConfigurationInstance.state().EnvironmentError).toEqual('\"value\" must be one of [context:EnvironmentOptions]');
		})
	});
});