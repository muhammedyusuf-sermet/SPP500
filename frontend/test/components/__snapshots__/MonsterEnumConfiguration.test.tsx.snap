// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster Creation Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Basic Configurations
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        xs={4}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Type"
          >
            Type
          </WithStyles(WithFormControlContext(InputLabel))>
          <MonsterDropdown
            disabled={false}
            name="Type"
            onChange={[Function]}
            options={
              Array [
                "Beast",
                "Dragon",
                "Monstrosity",
                "Humanoid",
                "Fiend",
                "Undead",
                "Giant",
                "Elemental",
                "SwarmOfTinyBeasts",
                "Construct",
                "Celestial",
                "Aberration",
                "Fey",
                "Plant",
                "Ooze",
              ]
            }
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Type-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={4}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Size"
          >
            Size
          </WithStyles(WithFormControlContext(InputLabel))>
          <MonsterDropdown
            disabled={false}
            name="Size"
            onChange={[Function]}
            options={
              Array [
                "Tiny",
                "Small",
                "Medium",
                "Large",
                "Huge",
                "Gargantuan",
              ]
            }
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Size-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={4}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Race"
          >
            Race
          </WithStyles(WithFormControlContext(InputLabel))>
          <MonsterDropdown
            disabled={false}
            name="Race"
            onChange={[Function]}
            options={
              Array [
                "AnyRace",
                "Devil",
                "Demon",
                "Human",
                "Shapechanger",
                "Goblinoid",
                "Titan",
                "Gnoll",
                "Gnome",
                "Dwarf",
                "Elf",
                "Orc",
                "Kobold",
                "Lizardfolk",
                "Merfolk",
                "Sahuagin",
                "Grimlock",
              ]
            }
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Race-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Alignment"
          >
            Alignment
          </WithStyles(WithFormControlContext(InputLabel))>
          <MonsterDropdown
            disabled={false}
            name="Alignment"
            onChange={[Function]}
            options={
              Array [
                "Unaligned",
                "AnyAlignment",
                "AnyNonGoodAlignment",
                "AnyNonEvilAlignment",
                "AnyNonLawfulAlignment",
                "AnyNonChaoticAlignment",
                "AnyGoodAlignment",
                "AnyEvilAlignment",
                "AnyLawfulAlignment",
                "AnyChaoticAlignment",
                "AnyNeutralAlignment",
                "LawfulGood",
                "LawfulNeutral",
                "LawfulEvil",
                "NeutralGood",
                "Neutral",
                "NeutralEvil",
                "ChaoticGood",
                "ChaoticNeutral",
                "ChaoticEvil",
              ]
            }
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Alignment-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Environment"
          >
            Environment
          </WithStyles(WithFormControlContext(InputLabel))>
          <MonsterDropdown
            disabled={false}
            name="Environment"
            onChange={[Function]}
            options={
              Array [
                "Arctic",
                "Coastal",
                "Desert",
                "Forest",
                "Grassland",
                "Hill",
                "Mountain",
                "Swamp",
                "Underdark",
                "Underwater",
                "Urban",
              ]
            }
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Environment-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
